scripts/scratch_ingest/expression_writer.py

```diff
@@ -0,0 +1,416 @@
+                    seek_point += 1
+                current_seek.append(seek_point)
+                seek_points.append(current_seek)
+                current_pos = seek_point + 1
+                current_seek = [current_pos]
+        return seek_points
+
+    @staticmethod
+    def read_sparse_matrix_slice(matrix_file_path, genes, data_dir, indexes):
+        """
+        Read a sparse matrix using start/stop indexes and append to individual gene-level files
+
+        :param matrix_file_path: (String): path to matrix file
+        :param genes: (List) gene names from features file
+        :param data_dir: (String) name out output dir
+        :param indexes: (List) start/stop index points to read from/to
+        """
+        start_pos, end_pos = indexes
+        print(f"reading {matrix_file_path} at index {start_pos}:{end_pos}")
+        with ExpressionWriter.open_file(matrix_file_path) as matrix_file:
+            current_pos = start_pos
+            matrix_file.seek(current_pos)
+            while current_pos < end_pos:
+                line = matrix_file.readline()
+                gene_idx = int(line.split()[0])
+                gene_name = genes[gene_idx - 1]
+                fragment_path = f"{data_dir}/gene_entries/{gene_name}__entries.txt"
+                with open(fragment_path, 'a+') as file:
+                    file.write(line)
+                current_pos += len(line)
+
+    def divide_sparse_matrix(self, genes, data_dir):
+        """
+        Slice a sparse matrix into 1GB chunks and write out individual
+        gene-level files in parallel
+
+        :param genes: (List) gene names from features file
+        :param data_dir: (String) name out output dir
+        """
+        print(f"loading sparse data from {self.matrix_file_path}")
+        slice_indexes = self.get_file_seek_points()
+        pool = multiprocessing.Pool(self.num_cores)
+        processor = partial(self.read_sparse_matrix_slice, self.matrix_file_path, genes, data_dir)
+        pool.map(processor, slice_indexes)
+
+    def process_fragment(self, barcodes, cluster_cells, cluster_name, data_dir, fragment_name):
+        """
+        Process a single-gene sparse matrix fragment and write expression array
+
+        :param barcodes: (List) list of cell barcodes
+        :param cluster_cells: (List) list of cells from cluster file
+        :param cluster_name: (String) name of cluster object
+        :param data_dir: (String) name out output dir
+        :param fragment_name: (String) name of gene-level fragment file
+        """
+        gene_name = fragment_name.split('__')[0]
+        full_path = f"{data_dir}/gene_entries/{fragment_name}"
+        observed_cells = []
+        exp_vals = []
+        # write array of 0 values if file is empty
+        if os.stat(full_path).st_size == 0:
+            empty_exp = [0] * len(cluster_cells)
+            ExpressionWriter.write_gene_scores(cluster_name, gene_name, empty_exp, data_dir)
```

Comment:
This is another area ripe for optimization, likely worth noting as a TODO.

You mentioned roughly 4k of 28k genes had no expression in our pilot dataset.  So we could write a list of gene symbols that lack expression in a special file, then background-read that into the frontend upon loading a cluster, and display a “0” gene expression scatter plot for all of them.  Then we could skip writing these thousands of files of all zeros.

That'd save sizable pipeline run time, end-user / Image Pipeline download and parse time, and storage cost.